### PR TITLE
feat: add knip CI configuration

### LINKED ISSUE #12: Knip CI
I know that we have knip ci on one of our other repositories. @gitcoindev perhaps you can add that in here?

### PR DESCRIPTION
**Pull request by gitcoindev**
Resolves: https://github.com/ubiquity/ts-template/issues/12

### CODE REVIEW

**Review by 0x4007 — COMMENTED** [review]
This is functional right? Last time I tried it, it didn't work. Can you test?

**Comment by gitcoindev** [comment]
Hi @pavlovcik I ported knip CI almost 1:1 from ubiquibot. Integration looks good , but failing with 
```
  - Create check ID
Error: 🧨 Failed: Resource not accessible by integration
Error: 📚 Stack: HttpError: Resource not accessible by integration
```
this is due to lack of permission in the `ts-template` repository to add pull request comments from GitHub actions. If I remember correctly we saw it already. The permission is required by Codex-/knip-reporter@v2 and the error can be resolved by comparing settings on bot repository and enabling write comment access.

**Comment by gitcoindev** [comment]
Yes, I can confirm this works on my fork with write settings enabled. QA: https://github.com/gitcoindev/ts-template/pull/1
Currently knip reports two unused dev dependencies: https://github.com/gitcoindev/ts-template/pull/1#issuecomment-1953201038

**Comment by 0x4007** [comment]
> comparing settings on bot repository and enabling write comment access.

Can you elaborate? Would like to fix this!

**Comment by gitcoindev** [comment]
> > comparing settings on bot repository and enabling write comment access.
> 
> Can you elaborate? Would like to fix this!

Yes sure,

Settings->Actions>General>Workflow permissions->Select 'Read and Write permissions'.

Choose the default permissions granted to the GITHUB_TOKEN when running workflows in this repository. You can specify more granular permissions in the workflow using YAML. [Learn more about managing permissions.](https://docs.github.com/actions/reference/authentication-in-a-workflow#modifying-the-permissions-for-the-github_token)



Read and write permissions

**Comment by 0x4007** [comment]
Perhaps we can auth it with our bot? This way we don't have to manually change the settings for all our repositories. 

All the values necessary are in our org secrets! Check out my logic for posting the comment as the bot in @ubiquity/cloudflare-deploy-action!

File: .github/workflows/knip.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+jobs:
+  run-knip:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Setup Node
+        uses: actions/setup-node@v3
```
Comment (0x4007) [inline .github/workflows/knip.yml]:
```suggestion
        uses: actions/setup-node@v4
```

File: .github/workflows/knip.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  pull_request:
+
+permissions: write-all
+
+jobs:
+  run-knip:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
```
Comment (0x4007) [inline .github/workflows/knip.yml]:
```suggestion
        uses: actions/checkout@v4
```

File: .github/workflows/knip.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+      - name: Setup Node
+        uses: actions/setup-node@v3
+        with:
+          node-version: 20.10.x
```
Comment (0x4007) [inline .github/workflows/knip.yml]:
```suggestion
          node-version: 20.10.0
```

**Review by 0x4007 — CHANGES_REQUESTED** [review]

File: .github/workflows/knip.yml (hunk 3 above)
Comment (gitcoindev) [reply .github/workflows/knip.yml]: done

File: .github/workflows/knip.yml (hunk 2 above)
Comment (gitcoindev) [reply .github/workflows/knip.yml]: done

File: .github/workflows/knip.yml (hunk 1 above)
Comment (gitcoindev) [reply .github/workflows/knip.yml]: done

**Review by 0x4007 — APPROVED** [review]

**Comment by gitcoindev** [comment]
> Perhaps we can auth it with our bot? This way we don't have to manually change the settings for all our repositories.
> 
> All the values necessary are in our org secrets! Check out my logic for posting the comment as the bot in @ubiquity/cloudflare-deploy-action!

An interesting idea. I will try to experiment on that and try to submit a solution till today eod.

**Comment by 0x4007** [comment]
Use this as inspiration. The same exact property names should work for any repository under @ubiquity. 

https://github.com/ubiquity/cloudflare-deploy-action/blob/main/src/handle-commit.ts

**Pull request merged**